### PR TITLE
Cascaded trainer architecture

### PR DESCRIPTION
**Pull request by rahul1980**
A cascade is defined as a set of delegates and each final transition on a parser state is determined by a sequence of delegates. Each delegate either decides on a part (or all) of the parser action, or delegates the decision to another delegate. An example of such a cascade is the ShiftPropbankEvoke cascade, where the first delegate decides whether to SHIFT or not, a second one decides (in case of a non-SHIFT action) whether to output a non-Propbank-EVOKE action or not, and a third one decides (in case of a Propbank EVOKE) on the specific PropBank EVOKE action.

This is currently modeled as one softmax layer per delegate, and all softmax layers are connected to the feedforward unit's hidden layer. These layers are individually significantly thinner than the current monolithic softmax layer.

The cascade architecture offers the following advantages:
a) Speedup: We only invoke the softmax layers of delegates as required, and these layers are smaller. E.g. roughly half of our actions are SHIFT, which means that half the time we don't compute anything beyond the binary output ShiftOrNot delegate.

b) Expressiveness: We can add and pair custom delegates as required. Each delegate can define its own loss, e.g. ranking loss instead of softmax loss for coreference delegates. The proposed architecture goes beyond just softmax layers/losses.

c) Scalability: As we add more action types and instances, the current monolithic architecture becomes increasingly untenable. Having cascades allows us to refactor the modeling into small delegates.

This PR also makes the ShiftPropbankEvoke cascade the default for CASPAR. Without any optimizations, the forward pass with this cascade is 2.2x faster, and the backward pass is roughly 1.2x faster, leading to an overall 25% increasing in training speed for the same accuracy.

### CODE REVIEW

File: sling/nlp/parser/trainer/pytorch_modules.py [hunk 1]
```diff
@@ -385,105 +471,75 @@ def forward(self, document, train=False):
...
+        # Store the last CASCADE action in a cascade.
+        last = None
+        while cascading:
+          # Get the highest scoring action from the cascade delegate.
+          # Note: We don't have to do any filtering or checking here, we
+          # can just return the top-scoring action.
+          best = cascade.predict(delegate_index, state, last, ff_hidden)
+
+          if best.is_cascade():
+            delegate_index = cascade.next(best, delegate_index)
```
Comment (ringgaard) [inline sling/nlp/parser/trainer/pytorch_modules.py]: Why isn't the delegate index not just encoded in the action? Here you instead call the cascade to determine the next delegate index.

File: sling/nlp/parser/tools/train_pytorch.py [hunk 2]
```diff
@@ -182,22 +183,25 @@ def clip_gradients(self):
...
+      self.last_update_count = self.count
+      print self.batch_losses.tostring(self.count)
       start = time.time()
-      self.batch_loss /= self.current_batch_num_transitions
 
-      # Add the regularization penalty to the batch loss.
+      objective = self.batch_losses.average()
+
+      # Add the regularization penalty to the objective.
       l2 = Var(torch.Tensor([0.0]))
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/tools/train_pytorch.py]: I am printing the l2 penalty a few lines below, so I need it declared here.

File: sling/nlp/parser/trainer/pytorch_modules.py [hunk 3]
```diff
@@ -210,12 +287,21 @@ def __init__(self, spec):
       self.add_module('ff_link_transform_' + f.name, transform)
       f.transform = transform
 
-    # Feedforward unit.
+    # Feedforward unit trunk.
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/trainer/pytorch_modules.py]: Now that the outputs (aka heads) of the FF unit are split off into their own cells, what remains here is the trunk. I have seen the same nomenclature in codebases for similar architecture at multiple places.

File: sling/nlp/parser/trainer/cascade.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+  def index(self, action):
+    pass
+
+  """Returns the action corresponding to 'index' in the softmax layer."""
+  def action(index, previous_action=None):
+    pass
+
+  """Computes the delegate loss, where 'ff_hidden' is the output of the FF unit's
+  hidden layer, and 'gold' is the gold transition for the delegate."""
+  def loss(self, state, ff_hidden, gold):
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/trainer/cascade.py]: Changed to activation (and ff_activation where relevant).

File: sling/nlp/parser/trainer/pytorch_modules.py [hunk 5]
```diff
@@ -20,9 +20,13 @@
 import torch
 import torch.nn as nn
 
+sys.path.insert(0, "sling/nlp/parser/trainer")
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/trainer/pytorch_modules.py]: removed.

File: sling/nlp/parser/trainer/pytorch_modules.py [hunk 6]
```diff
@@ -181,6 +185,79 @@ def __repr__(self):
...
+    return s
+
+
+# Asserts 'delegate' to be a softmax delegate.
+def assert_softmax_delegate(delegate):
+  assert isinstance(delegate, SoftmaxDelegate), delegate.__class__.__name__
+
+
 # Top-level module for SEMPAR.
 class Sempar(nn.Module):
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/trainer/pytorch_modules.py]: Done.

File: sling/nlp/parser/trainer/pytorch_modules.py (hunk 1 above)
Comment (rahul1980) [reply sling/nlp/parser/trainer/pytorch_modules.py]: Done.

**Comment by rahul1980** [comment]
Michael,  I think I have addressed your comments. I have also added the cascade-aware parser runtime to this PR since it is a relatively small change and this way all cascade changes are together in this PR.
The runtime changes are in the two latest commits, where the first commit has most of the changes. You can view that specific commit here:
https://github.com/google/sling/pull/213/commits/a2ca0166c0f11c8114455e81d26451e02bb95446

I have tested the runtime by training and evaluating on multiple small corpora (10 and 50 documents) and comparing against the python evaluator.

I also took the opportunity to remove some unused python code from the trainer.

**Comment by rahul1980** [comment]
Thanks Michael, I think I have addressed your comments. The cascade code is now in cascade.{h, cc} and there is a DelegateInstance class now that wraps the myelin instance for the delegate cell. PTAL.

**Comment by ringgaard** [comment]
I am not sure we got the separation of matters in the cascade/delegate setup quite right yet.

 The Cascade and CascadeInstance classes are part of the parser and there is only one implementation of these classes. They form the upper part of the three-layer structure of the general parser:
  1) Lexical encoder
  2) "Trunk"
  3) Cascade

The Delegate and DelegateInstances are the component interfaces (defined in parser.h also) to interface with the different delegate component types. They can contain arbitrary models for predicting the next action, and the parser should not make more assumptions than needed about the internals of these. I know that we only have the "softmax" delegates for now. I am not sure "softmax" is the most descriptive name for these. They are more characterized by using a simple linear layer for predicting among a fixed set of actions. 

The parser needs to be able to do the following:

1) Create and initialize a Cascade. The Cascade will create and initialize each of the delegates using the spec. This is done when the parser is initialized. 
2) Create a CascadeInstance from an intialized Cascade object. This will call each delegate to return a DelegateInstance for the delegate. This is done each time Parse(doc) is called.
3) At parsing time, we call instance.cascade->Predict(activation, &action) to get the next action. This will start with the main delegate and call instance.cascade.delegate(activation, &action) to allow the delegate to predict the next action. If this action is a CASCADE action, we loop until we have the final action.
 
Maybe this requires a face-to-face discussion. I am fine with you submitting this as-is now, and then we can discuss this in more details when I am back from vacation.

**Comment by rahul1980** [comment]
Thanks Michael, I am going to submit as-is and we can reorganize this once you are back.
